Implementation of Yen's K shortest path algorithm.

kevemueller opened the pull request:
Features the actual algorithm implementation and a test framework to run
multiple parameterized tests against different K shortest algorithms.

jkinable commented:
Thanks for your submission. We'll try to review this as soon as possible but we currently have a bit of a review backlog so it may take some time.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPaths.java:
```diff
@@ -0,0 +1,114 @@
+/*
+ * (C) Copyright 2017, by Keve Müller.
```
Please use the original copyright statement as included in all files. The first copyright line should read:
* (C) Copyright 2017-2017, by Keve Müller and Contributors.

We use automated scripts to update the years, and these scripts don't like deviations in the copyright statement...

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPaths.java:
```diff
@@ -0,0 +1,114 @@
+ * edge cost. The algorithm was published by Jin Y. Yen in 1971 and employs any shortest path
+ * algorithm to find the best path, then proceeds to find K-1 deviations of the best path.
+ * <p>
+ * Based on pseudocode published at https://en.wikipedia.org/wiki/Yen%27s_algorithm
```
Please state the runtime complexity of your implementation using big-O notation:
$O(runtime)$
Use the $$ signs.
See "javadoc and comment" and "math in javadoc": https://github.com/jgrapht/jgrapht/wiki/Contributor-Guidelines

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPathsIterator.java:
```diff
@@ -0,0 +1,283 @@
+                spFactory.apply(maskedSubgraph);
+            GraphPath<V, E> spurPath = maskedShortestSpurPathAlgorithm.getPath(spurNode, sink);
+
+            if (null != spurPath && 0 != spurPath.getLength()) { // BellmannFord
```
BellmanFord is (by contract) required to return null if no path exists. This is sounds like a bug. Could you indicate where this occurs?

kevemueller replied:
Just take a look at the result of org.jgrapht.alg.shortestpath.MultiShortestPathsStringDWETest
[YenKShortestPaths+BellmanFord/Eppstein_11-0]
BellmannFord returns an empty path.

d-michail replied:
Opened #433 to resolve this. Adjust accordingly after merge.

kevemueller replied:
Removed the special handling of Bellman after #433 was merged.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPathsIterator.java:
```diff
@@ -0,0 +1,283 @@
+                                                                 // instead
+                                                                 // of null
+                // Entire path is made up of the root path and spur path.
+                ArrayList<E> totalPath = new ArrayList<E>();
```
This is rather inefficient. You could do the following:
1. you have already calculated the rootPath which is stored as a list of vertices List<V> rootPath.  Consequently you also know the cost of the rootPath C_rootpath
2. You have calculated the spurPath as a GraphPath with cost C_spurPath

You could now simply create a new GraphWalk as follows:
```
List<V> totalPath=new ArrayList<>(rootPath.subList(0,rootPath.size()-1));
totalPath.addAll(spurPath.getVertexList);
GraphWalk<V, E> gw = new GraphWalk<V, E>(graph, source, target, totalPath, C_rootPath+C_spurPath);
```
List<V> totalPath=new ArrayList<>(rootPath.subList(0,rootPath.size()-1));
totalPath.addAll(spurPath.getVertexList);
GraphWalk<V, E> gw = new GraphWalk<V, E>(graph, source, target, totalPath, C_rootPath+C_spurPath);

It is technically quite annoying that the ShortestPath algorithms return a GraphPath instead of a GraphWalk, otherwise you could have simply used the concat function from GraphWalk... We should probably address this.

kevemueller replied:
I have tried your suggestion. It is temptingly elegant, but unfortunately not correct. After adding the needed ifs the code looked way uglier than the simple iteration that is in my initial code.
I kept the working code.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPathsIterator.java:
```diff
@@ -0,0 +1,283 @@
+            if (null == csp) {
+                return;
+            }
+        } while (lastShortestPath.getEdgeList().equals(csp.getEdgeList()));
```
This confuses me. By definition list b should not contain paths which are identical to lastShortestPath, so this loop seems unnecessary? `a.add(b.poll())` should suffice?

kevemueller replied:
the null check was indeed redundant. removed it.
the loop is needed.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPathsIterator.java:
```diff
@@ -0,0 +1,283 @@
+        // node in the previous k-shortest path.
+        for (int i = 0; i < lastShortestPath.getLength() - 1; i++) {
+            // Spur node is retrieved from the previous k-shortest path
+            V spurNode = lastShortestPath.getVertexList().get(i);
```
This unfortunately is a bit tricky. A Path in a graph can be unambiguously expressed in terms of a sequence of edges. A path in a graph expressed by a sequence of vertices may however be ambiguous. For instance, in a multigraph where there may be multiple edges between the same pair of vertices, a path must be expressed in terms of the edges. Therefore, most shortest path algorithms will return a GraphPath expressed in terms of the edges of the path. Each invocation to getVertexList() may recompute the path expressed in terms of vertices. This gets potentially very expensive, since you invoke getVertexList() very often. It would be better to cache the result of this method. To understand what I mean, have a look at the implementation of GraphPath and GraphWalk.

kevemueller replied:
This is not ambiguous for Yen's algorithm, as Yen's algorithm will not work if there are multiple edges between the same vertices.

d-michail replied:
Indeed you need to use some caching here as method `getVertexList()` might reconstruct a new list on each invocation. Take a look at the internals of `GraphWalk` to understand this.

kevemueller replied:
I am storing now both the GraphPath as well as the call to getVertexList. IMO this is BAD DESIGN - it fixes a problem at the wrong place. 
If an implementer of GraphPath considers its implementation of getVertexList to be causing performance issues, that implementer of GraphPath should keep a weak reference to a once computed vertex list and return it in subsequent calls. 
Currently only GraphWalk implements GraphPath and it has already slots for both vertexList as well as edgeList, any caching logic should be added here.
Users of GraphPath should not be concerned with caching results.

jkinable replied:
I've been working on some optimizations for graphwalk, including this caching issue. Will be submitted soon.

hadronzoo commented:
Any update on this PR? This algorithm is very useful.

hadronzoo commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPathsIterator.java:
```diff
@@ -0,0 +1,283 @@
+     */
+    private void tryAddNext()
+    {
+        GraphPath<V, E> lastShortestPath = a.get(a.size() - 1);
```
This throws an exception if no paths are found.

kevemueller replied:
thx. indeed the most obvious case of no path was not handled properly. fixed.

kevemueller commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPaths.java:
```diff
@@ -0,0 +1,131 @@
+     * Construct the object for searching k shortest paths in provided graph. The shortest path
+     * algorithm defaults to {@link DijkstraShortestPath}.
+     * 
+     * @param graph the underlying graph on which the paths are searched
```
adapted.

kevemueller commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPaths.java:
```diff
@@ -0,0 +1,131 @@
+     * shortest path algorithm.
+     * 
+     * @param graph the underlying graph on which the paths are searched.
+     * @param hintK the hint on how many paths shall be returned by {@link #getPaths(Object, Object)}.
```
adapted.

kevemueller commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPaths.java:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    /**
+     * Get an iterator that returns the k-shortest paths from source to sink in increasing order.
```
adapted.

kevemueller commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPaths.java:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    @Override
+    public List<GraphPath<V, E>> getPaths(V source, V sink)
```
passing k in the constructor is indeed not nice.
As all k shortest path algorithms (so far) use an underlying iterator to produce the paths, I suggest to refactor as follows:
Declare an interface for ShortestPathIterator that can be used sub implement the different algorithms.
Use a generic k shortest path class that is call compatible with the current codebase and gets an ShortestPathIterator passed to retrieve the shortest paths.

kevemueller commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPaths.java:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    @Override
+    public List<GraphPath<V, E>> getPaths(V source, V sink)
```
adapted.

kevemueller commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPathsIterator.java:
```diff
@@ -0,0 +1,283 @@
+import org.jgrapht.graph.MaskSubgraph;
+
+/**
+ * Yen's algorithm computes single-source K-shortest loopless paths for a graph with non-negative
```
adapted.

kevemueller commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPathsIterator.java:
```diff
@@ -0,0 +1,283 @@
+ *
+ * @since June 1, 2017
+ */
+public final class YenKShortestPathsIterator<V, E>
```
no. please look at my suggestion above.

kevemueller commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenKShortestPathsIterator.java:
```diff
@@ -0,0 +1,283 @@
+        }
+        nextIndex = 0;
+        // // Initialize the heap to store the potential kth shortest paths.
+        b = new PriorityQueue<GraphPath<V, E>>(new Comparator<GraphPath<V, E>>()
```
adapted.

kevemueller commented:
@jkinable Thank you very much for the exhaustive review of the code. I have followed your advice to the extent possible and adapted most of your recommendations.
@hadronzoo Thank you very much for spotting the bug for the trivial case. I have added a test case and changed the code.
I hope this can be merged now...

kevemueller commented:
@jkinable On day 265 after the initial commit I just wanted to check with you if you consider merging this at all.

TomasJohansson commented:
It does not seem as YenKShortestPaths works properly.

Here is a failing test for a small graph with five edges and three expected paths:

https://github.com/TomasJohansson/adapters-shortest-paths/blob/daae43a2411fe9e644fcf0e29b5f3a9640912028/adapters-shortest-paths-impl-jgrapht_fork_kevemueller/src/test/java/jgrapht_fork_kevemueller/KShortestPathAlgorithmTest.java

pecollet commented:
> It does not seem as YenKShortestPaths works properly.
> 
> Here is a failing test for a small graph with five edges and three expected paths:
> 
> https://github.com/TomasJohansson/adapters-shortest-paths/blob/daae43a2411fe9e644fcf0e29b5f3a9640912028/adapters-shortest-paths-impl-jgrapht_fork_kevemueller/src/test/java/jgrapht_fork_kevemueller/KShortestPathAlgorithmTest.java

That test seems incorrect.
The third result of the path computation can be either A-B-C-D (21) or A-C-B-D (21), as both have the same weight.
The assertion expects A-B-C-D only.

TomasJohansson commented:
> > It does not seem as YenKShortestPaths works properly.
> > Here is a failing test for a small graph with five edges and three expected paths:
> > https://github.com/TomasJohansson/adapters-shortest-paths/blob/daae43a2411fe9e644fcf0e29b5f3a9640912028/adapters-shortest-paths-impl-jgrapht_fork_kevemueller/src/test/java/jgrapht_fork_kevemueller/KShortestPathAlgorithmTest.java
> 
> That test seems incorrect.
> The third result of the path computation can be either A-B-C-D (21) or A-C-B-D (21), as both have the same weight.
> The assertion expects A-B-C-D only.

@pecollet Please note the "Directed" part of the "SimpleDirectedWeightedGraph".
That semantic would be awkward if edges in both directions (e.g. also an edge from C to B) would be constructed with this code from the linked test case:
```java
graph = new SimpleDirectedWeightedGraph<String, DefaultWeightedEdge>(DefaultWeightedEdge.class);
Graphs.addEdgeWithVertices(graph, "A", "B", 5);
Graphs.addEdgeWithVertices(graph, "A", "C", 6);
Graphs.addEdgeWithVertices(graph, "B", "C", 7);
Graphs.addEdgeWithVertices(graph, "B", "D", 8);
Graphs.addEdgeWithVertices(graph, "C", "D", 9);
```
Also, the Yen test actually finds only the first two asserted paths (i.e. A-B-D and A-C-D) i.e. NONE of those two you mention (A-B-C-D or A-C-B-D )

pecollet commented:
> @pecollet Please note the "Directed" part of the "SimpleDirectedWeightedGraph".
I had missed that detail indeed. I was assuming non-directed edges.

Toptachamann commented:
Implemented in #764
